### PR TITLE
feat: Incremental Gradle builds for vaadinPrepareFrontendTask

### PR DESCRIPTION
**Pull request by mshabarov**
## Description

Defines the input and outputs for `vaadinPrepareFrontendTask` of Vaadin Gradle plugin making it possible to skip the task if inputs and outputs are not changed, giving `UP-TO-DATE` result, according to [incremental builds feature](https://docs.gradle.org/current/userguide/incremental_build.html#incremental_build)

Inputs are:
- path to Node.js executable
- path to npm/pnpm executable
- configuration parameters, defined in [VaadinFlowPluginExtension](https://github.com/vaadin/flow/blob/main/flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinFlowPluginExtension.kt)

Outputs are:
- `package.json`, `vite.config.js` and other files generated by Vaadin in the root of the project and in the frontend folder
- frontend folder, including generated folder

according to a prepareFrontend definition here https://github.com/vaadin/flow/blob/04a69001101aec44cb43f190ca91b4b4d1491dcf/flow-plugins/flow-plugin-base/src/main/java/com/vaadin/flow/plugin/base/BuildFrontendUtil.java#L148

How to test:
- in Flow:
`git checkout cache-prepare-frontend`
`mvn clean install -DskipTests`
`cd flow-plugins/flow-gradle-plugin`
`./gradlew clean build publishToMavenLocal -x functionalTest`
- in base-starter-gradle:
add the following to `build.gradle`
```
buildscript {
    repositories {
        mavenLocal()
        ...
    }
}

repositories {
    mavenLocal()
    ...
}
```
add the following to `gradle.properties`
```
vaadinVersion=24.2-SNAPSHOT
```
run `./gradlew clean vaadinPrepareFrontend --info`

*Open questions*:
- currently there is no input or output corresponding to changes in classpath (e.g. when a JAR dependency is added, e.g. add-on or reused theme), so that the `TaskCopyFrontendFiles` might be skipped, because the whole prepare-frontend is skipped as it doesn't see changes in classpath. Not sure if it's an issue, because in dev mode this is triggered by `DevModeInitializer` and in prod mode with the `build-frontend`.
- probably it's better to remove `@CacheableTask` for V24.2 (replace it with `@DisableCachingByDefault`) and let developers enable caching manually. The way how to disable caching for a task is described [here](https://docs.gradle.org/7.6.1/userguide/more_about_tasks.html#sec:disable-state-tracking), but I couldn't make it work neither with `tasks.register` nor with `tasks.findByName` so far -> needs more testing and we have to suggest developers how to disable caching in case if Gradle false thinks that its up-to-date when it's not.
- no tests so far

Fixes https://github.com/vaadin/flow/issues/15342

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [x] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinPrepareFrontendTask.kt [hunk 1]
```diff
@@ -18,20 +18,45 @@ package com.vaadin.gradle
...
 public open class VaadinPrepareFrontendTask : DefaultTask() {
 
+    private val extension: VaadinFlowPluginExtension
+
+    /**
+     * Defines an object containing all the inputs of this task.
+     */
+    @Nested
+    public open fun getTaskInputProperties(): PrepareFrontendInputProperties? {
+        return PrepareFrontendInputProperties(project)
```
Comment (mcollovati) [inline flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinPrepareFrontendTask.kt:44]: The `getTaskInputProperties()` method is called multiple times. We can perhaps introduce a `PrepareFrontendInputProperties` field and instantiate it in the `init` block.

File: flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinPrepareFrontendTask.kt [hunk 2]
```diff
@@ -18,20 +18,45 @@ package com.vaadin.gradle
...
+    }
+
+    /**
+     * Defines an object containing all the outputs of this task.
+     */
+    @Nested
+    public open fun getTaskOutputProperties(): PrepareFrontendOutputProperties? {
+        return PrepareFrontendOutputProperties(project)
```
Comment (mcollovati) [inline flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinPrepareFrontendTask.kt:52]: The `getTaskOutputProperties()` method is called multiple times. We can perhaps introduce a `PrepareFrontendOutputProperties` field and instantiate it in the `init` block.

File: flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/PrepareFrontendOutputProperties.kt [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    @OutputDirectory
+    public fun getFrontendDirectory(): File {
+        return extension.frontendDirectory
+    }
+
+    @OutputDirectory
+    public fun getGeneratedTsFolder(): File {
+        return extension.generatedTsFolder
+    }
+}
```
Comment (mcollovati) [inline flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/PrepareFrontendOutputProperties.kt:86]: newline missing
```suggestion
}

```

File: flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/PrepareFrontendInputProperties.kt [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+            FrontendUtils.getVaadinHomeDirectory()
+                .absolutePath
+        }
+        settings.nodeDownloadRoot = URI(extension.nodeDownloadRoot)
+        settings.isForceAlternativeNode = extension.requireHomeNodeExec
+        settings.isUseGlobalPnpm = extension.useGlobalPnpm
+        settings.isAutoUpdate = extension.nodeAutoUpdate
+        return FrontendTools(settings)
+    }
+}
```
Comment (mcollovati) [inline flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/PrepareFrontendInputProperties.kt:259]: newline missing 
```suggestion
}

```

**Review by mcollovati — CHANGES_REQUESTED** [review]

**Comment by mcollovati** [comment]
> currently there is no input or output corresponding to changes in classpath (e.g. when a JAR dependency is added, e.g. add-on or reused theme), so that the TaskCopyFrontendFiles might be skipped, because the whole prepare-frontend is skipped as it doesn't see changes in classpath. Not sure if it's an issue, because in dev mode this is triggered by DevModeInitializer and in prod mode with the build-frontend.

The `GradlePluginAdapter` has a `getJarFiles()` method that could probably be used as `@Input` to track dependencies changes. But, as commented earlier, it may not be worth it, since the `TaskCopyFrontendFiles` is executed anyway. 

> probably it's better to remove @CacheableTask for V24.2 (replace it with @DisableCachingByDefault) and let developers enable caching manually. The way how to disable caching for a task is described [here](https://docs.gradle.org/7.6.1/userguide/more_about_tasks.html#sec:disable-state-tracking), but I couldn't make it work neither with tasks.register nor with tasks.findByName so far -> needs more testing and we have to suggest developers how to disable caching in case if Gradle false thinks that its up-to-date when it's not.

It seems to me that the cache annotations are not influencing the `up-to-date` state.
An alternative may be to use `doNotTrackState()` in the plugin `init` block, but make the execution conditional based on a new plugin property, e.g. `disableVaadinPrepareFrontendTaskTrackState` with default `true`.

```
init {
        if (extension.disableVaadinPrepareFrontendTaskTrackState) {
            doNotTrackState("State tracking for vaadinPrepareFronted task is disabled (default). Use the 'disableVaadinPrepareFrontendTaskTrackState' plugin property to enable the feature");
        }
}
```

This way, the new feature will not be active by default, but the developer can opt-in by setting the property in the `vaadin` section of the gradle project file.

```
vaadin {
    pnpmEnable = false
    disableVaadinPrepareFrontendTaskTrackState = false
}
```

In future releases, we can change the default value to always activate the feature active and then deprecate and remove the setting (or leave it there, it can be useful).

File: flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/PrepareFrontendInputProperties.kt [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    }
+
+    @Input
+    @Optional
+    public fun getResourceOutputDirectory(): String? {
+        val resourceOutputDirectory = extension.resourceOutputDirectory
+        if (!resourceOutputDirectory.exists()) {
+            return null
+        }
+        return resourceOutputDirectory.absolutePath
```
Comment (mcollovati) [inline flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/PrepareFrontendInputProperties.kt:186]: The prepareFrontend task generates the `flow-build-info.json` file in this directory.
This causes the second run of the task to be considered not up-to-date.
I wonder if this directory should instead be considered a task output

**Comment by mcollovati** [comment]
Also `vaadin-featureflags.properties` should be considered, since `frontend/generated/vaadin-featureflags.js` is generated and updated by the prepare frontend task, based on the contents of the file

File: flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/PrepareFrontendInputProperties.kt (hunk 5 above)
Comment (mshabarov) [reply flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/PrepareFrontendInputProperties.kt:186]: Using a local `TestProject()` object within the test method should work I suppose.

Comment (mcollovati) [reply flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/PrepareFrontendInputProperties.kt:186]: My understanding was that the temporary directory is consistent in the same test, unless you create a new instance of `TestProject`

Comment (mshabarov) [reply flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/PrepareFrontendInputProperties.kt:186]: Right, should be the same within the same test, weird.

Comment (mshabarov) [reply flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/PrepareFrontendInputProperties.kt]: I think you were right about moving resource dir property to an output. If we do so, Gradle would still be able to spot the re-configuring the path change and trigger a new build.

File: flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinPrepareFrontendTask.kt (hunk 1 above)
Comment (mshabarov) [reply flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinPrepareFrontendTask.kt]: done

File: flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinPrepareFrontendTask.kt (hunk 2 above)
Comment (mshabarov) [reply flow-plugins/flow-gradle-plugin/src/main/kotlin/com/vaadin/gradle/VaadinPrepareFrontendTask.kt]: done

**Review by mcollovati — APPROVED** [review]

**Pull request merged**